design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+  Map<String, Map<Integer, Long>> localLagMap = request.getLag();
+  globalLagMap.put(server, localLagMap);
+}
+``` 
+
+### Lag-aware routing
+
+Given the above information (alive + lag), how does a KSQL server decide to which standby to route 
+the request? Every server knows the lag of all partitions hosted by all servers in the cluster. 
+This allows us to implement lag-aware routing where server `B` can a) determine that server `A` is 
+down and b) decide whether it will serve the request itself or forward it to `C` depending on who 
+has the smallest lag for the given key in the pull query. 
+
+Pseudocode for routing:
+```java
+// Map populated periodically through heartbeat information
+Map<KsqlNode, Boolean> aliveNodes;
+
+// Map populated periodically through lag reporting
+Map<KsqlNode, Map<String, Map<Integer, Long>> globalLagMap;
+ 
+// Fetch the metadata related to the key
+KeyQueryMetadata queryMetadata = queryMetadataForKey(store, key, serializer);
+ 
+// Acceptable lag for the query
+final long acceptableOffsetLag = 10000;
+
+// Ordered list of servers, in order of most caught-up to least
+List<KsqlNode> nodesToQuery;
+KsqlNode active = queryMetadata.getActiveHost();
+if (aliveNodes.get(active)) {
+  nodesToQuery.add(active)  
+}
+
+// add standbys
+nodesToQuery.addAll(queryMetadata.getStandbyHosts().stream()
+    // filter out all the standbys that are down
+    .filter(standbyHost -> aliveNodes.get(standByHost) == null)
+    // get the lag at each standby host for the key's store partition
+    .map(standbyHost -> new Pair(standbyHostInfo, globalLagMap.get(standbyHost).get(storeName).get(queryMetadata.partition())))
+    // Sort by offset lag, i.e smallest lag first
+    .sorted(Comaparator.comparing(Pair::getRight())
+    .filter(standbyHostLagPair -> standbyHostLagPair.getRight() < acceptableOffsetLag)
+    .map(standbyHostLagPair -> standbyHostLagPair.getLeft())
+    .collect(Collectors.toList()));
+ 
+// Query available nodes, starting from active if up
+List<Row> result = null;
+for (KsqlNode server : nodesToQuery) {
+  try {
+    result = query(store, key, server);
+  } catch (Exception e) {
+    System.err.println("Querying server %s failed", server);
+  }
+  if (result != null) {
+    break;
+  }
+}
+
+if (result == null) {
+  throw new Exception("Unable to serve request. All nodes are down or too far behind");
+}
+
+return result;
+```
+
+We will also introduce a configuration `ksql.query.pull.acceptable.offset.lag`, that will provide 
+applications the ability to control how much stale data it is willing to tolerate, for sake of availability. 
+If a standby lags behind by more than the tolerable limit, pull queries will fail. This is a very 
+useful knob to handle the scenario of cold start explained above. In such a case, a newly added KSQL 
+server could be lagging by a lot as it rebuilds the entire state and thus the usefulness of the data 
+returned by pull queries may diminish significantly.
```

Comment:
Whether a server is ready to serve queries is determined by the lag. If the server lags too far behind or if there are others with smaller lag, then it will not serve queries. The heartbeat mechanism just provides information on whether a server is up or down. We need to get the actual lag information, to make further decisions. What does it mean to be caught up at least once? For the lag to have been zero at some point in time? What does that achieve?

Reply:
> What does it mean to be caught up at least once? For the lag to have been zero at some point in time?

Yes that's what it means. When I wrote this I had thought of a few good use cases why we might want it but my mind is GC'ing right now and can't come up with anything... I'll loop back if I remember but if not we can drop this